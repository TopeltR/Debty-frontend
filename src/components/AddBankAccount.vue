<template>
    <b-modal ref="modal" class="col-12">
        <div class="col-12" slot="modal-title">
            Add bank account
        </div>
        <div class="col-10 offset-1">
            <b-row>
                <b-col cols="12" class='TM30px'>
                    <b-row>
                        <form @submit.prevent='' class='wide'>
                            <div class='form-group'>
                                <label for='bankAccountName'>Name connected to bank account:</label>
                                <input type='text' class='form-control' v-model='bankAccount.name' id='bankAccountName'
                                       required placeholder='John Doe' maxlength="255">
                            </div>
                            <div class='form-group'>
                                <label for='bankAccountNumber'>Bank account number:</label>
                                <input type='text' class='form-control' v-model='bankAccount.number' maxlength="255"
                                       id='bankAccountNumber' required placeholder='EE1111111111111111'>
                            </div>
                        </form>
                    </b-row>
                </b-col>
            </b-row>
        </div>
        <div class="col-12" slot="modal-footer">
            <b-row>
                <b-col cols="6">
                    <b-button class="wide" variant="secondary" v-on:click="cancel">Cancel</b-button>
                </b-col>
                <b-col cols="6">
                    <b-button class="wide" variant="primary" v-on:click="saveBankAccount">Save</b-button>
                </b-col>
            </b-row>
        </div>
    </b-modal>
</template>

<script>
    import userStore from '@/stores/UserStore';

    export default {
        name: 'AddBankAccount',
        props: {
            state: {
                type: Object,
                showing: {
                    type: Boolean,
                },
                default: () => ({
                    showing: false,
                }),
            },
            eventId: {
                type: Number,
            },
        },
        watch: {
            state: {
                deep: true,
                handler(state) {
                    if (state.showing) {
                        this.$refs.modal.show();
                    } else {
                        this.$refs.modal.hide();
                    }
                },
            },
        },
        data: () => ({
            bankAccount: {
                name: undefined,
                number: undefined,
            },
        }),
        methods: {
            async saveBankAccount() {
                const user = await userStore.getUser();
                const result = await this.$http.post('/users/bankAccount/' + user.id, this.bankAccount);
                this.cancel();
            },
            cancel() {
                this.state.showing = false;
            },
        },
    };
</script>

<style scoped>
    .wide {
        width: 100%;
    }
</style>
