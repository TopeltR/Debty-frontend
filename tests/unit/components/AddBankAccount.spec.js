
import AddBankAccount from '@/components/AddBankAccount';
import {mount} from "@vue/test-utils";
import userStore from '@/stores/UserStore'
import {Vue} from "vue-property-decorator";
import BootstrapVue from 'bootstrap-vue';


describe('AddBankAccount.vue', () => {
    Vue.use(BootstrapVue);

    const bankAccount = {name: 'Name', number: 123};
    userStore.getUser = jest.fn(() => Promise.resolve({id: 1}));

    describe('saveBankAccount', () => {
        it('should make post request to address /users/bankAccount/{id} with bankAccount and call cancel', async () => {
            const addBankAccount = mount(AddBankAccount);
            const cancel = jest.spyOn(addBankAccount.vm, 'cancel');
            const post = jest.fn(() => Promise.resolve());
            addBankAccount.vm.$http = {post};
            addBankAccount.vm.bankAccount = bankAccount;

            await addBankAccount.vm.saveBankAccount();

            expect(post).toHaveBeenCalledWith('/users/bankAccount/1', bankAccount);
            expect(cancel).toHaveBeenCalled();
        })
    });

    const mockModal = (addBankAccount) => {
        const show = jest.fn();
        const hide = jest.fn();
        addBankAccount.vm.$refs.modal = {show, hide};
        return {show, hide};
    };

    describe('cancel', () => {
        it('should set state.showing to false', () => {
            const addBankAccount = mount(AddBankAccount);
            mockModal(addBankAccount);
            addBankAccount.vm.state.showing = true;
            addBankAccount.vm.cancel();

            expect(addBankAccount.vm.state.showing).toBe(false);
        })
    });

    describe('watch', () => {
        const addBankAccount = mount(AddBankAccount);
        const {show, hide} = mockModal(addBankAccount);
        it('should call modal.show when state.showing changes to true', () => {
            addBankAccount.vm.state.showing = false;
            addBankAccount.vm.state.showing = true;
            expect(show).toHaveBeenCalled();
        });

        it('should call modal.hide when state.showing changes to false', () => {
            addBankAccount.vm.state.showing = true;
            addBankAccount.vm.state.showing = false;
            expect(hide).toHaveBeenCalled();
        })
    });

    it('should render', () => {
        expect(mount(AddBankAccount).html()).toMatchSnapshot();
    })
});
