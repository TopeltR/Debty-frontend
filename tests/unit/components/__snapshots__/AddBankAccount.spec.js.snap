// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddBankAccount.vue should render 1`] = `
<div class="col-12">
  <div style="position: absolute; z-index: 2000;">
    <div role="dialog" tabindex="-1" aria-hidden="true" class="modal fade" style="display: none;">
      <div class="modal-dialog modal-md">
        <div role="document" class="modal-content">
          <header class="modal-header">
            <h5 class="modal-title">
              <div class="col-12">
                Add bank account
              </div>
            </h5><button type="button" aria-label="Close" class="close">Ã—</button>
          </header>
          <div class="modal-body">
            <div class="col-10 offset-1">
              <div class="row">
                <div class="TM30px col-12">
                  <div class="row">
                    <form class="wide">
                      <div class="form-group"><label for="bankAccountName">Name connected to bank account:</label> <input type="text" id="bankAccountName" required="required" placeholder="John Doe" class="form-control"></div>
                      <div class="form-group"><label for="bankAccountNumber">Bank account number:</label> <input type="text" id="bankAccountNumber" required="required" placeholder="EE1111111111111111" class="form-control"></div>
                    </form>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <footer class="modal-footer">
            <div class="col-12">
              <div class="row">
                <div class="col-6"><button type="button" class="btn wide btn-secondary">Cancel</button></div>
                <div class="col-6"><button type="button" class="btn wide btn-primary">Save</button></div>
              </div>
            </div>
          </footer>
        </div>
      </div>
    </div>
    <!---->
    <!---->
  </div>
</div>
`;
